### PR TITLE
Feat/int 1608 category exclusion

### PR DESCRIPTION
**Pull request by joyet-simon**

### CODE REVIEW

**Review by ClaireAlma — APPROVED** [review]

**Review by Benjamin-Freoua-Alma — APPROVED** [review]

File: cartridges/int_alma/cartridge/controllers/Cart.js [hunk 1]
```diff
@@ -19,12 +19,20 @@ server.append('Show', function (req, res, next) {
 server.get('BasketWidgetData', server.middleware.https, function (req, res, next) {
     var BasketMgr = require('dw/order/BasketMgr');
     var almaPlanHelper = require('*/cartridge/scripts/helpers/almaPlanHelper');
+    var almaHelpers = require('*/cartridge/scripts/helpers/almaHelpers');
+    var almaWidgetHelper = require('*/cartridge/scripts/helpers/almaWidgetHelper');
```
[the remark this answers is not in the document]
Comment (joyet-simon) [reply cartridges/int_alma/cartridge/controllers/Cart.js:23]: Yes, I need to include this in the route.

File: cartridges/int_alma/cartridge/controllers/Cart.js [hunk 2]
```diff
@@ -19,12 +19,20 @@ server.append('Show', function (req, res, next) {
...
 
     var currentBasket = BasketMgr.getCurrentBasket();
     var amount = Math.round(currentBasket.totalGrossPrice.multiply(100).value);
 
+    var productIds = [];
+    currentBasket.getAllProductLineItems().toArray().forEach(function (productLineItem) {
```
Comment (Francois-Gomis) [inline cartridges/int_alma/cartridge/controllers/Cart.js:29]: to array is necessary? Can you use object in collection ?

File: cartridges/int_alma/cartridge/controllers/Checkout.js [hunk 3]
```diff
@@ -43,13 +43,19 @@ server.append('Begin', function (req, res, next) {
...
+    var almaHelpers = require('*/cartridge/scripts/helpers/almaHelpers');
     var almaConfigInfo = getAlmaInfo();
 
     var BasketMgr = require('dw/order/BasketMgr');
     var currentBasket = BasketMgr.getCurrentBasket();
 
-    // if alma isn't activated don't even bother to call any API
-    if (!almaConfigInfo.isAlmaEnable) {
+    var productIds = [];
+    currentBasket.getAllProductLineItems().toArray().forEach(function (productLineItem) {
```
Comment (Francois-Gomis) [inline cartridges/int_alma/cartridge/controllers/Checkout.js:53]: to array is necessary? Can you use object in collection ?

File: cartridges/int_alma/cartridge/scripts/helpers/almaHelpers.js [hunk 4]
```diff
@@ -123,6 +123,43 @@ function getLocale(req) {
...
+ */
+function haveExcludedCategory(productIds) {
+    var ProductMgr = require('dw/catalog/ProductMgr');
+    var categoriesID = [];
+
+    productIds.forEach(function (productId) {
+        var product = ProductMgr.getProduct(productId);
+
+        if (product.isMaster()) {
+            product.getAllCategories().toArray().forEach(function (category) {
```
Comment (Francois-Gomis) [inline cartridges/int_alma/cartridge/scripts/helpers/almaHelpers.js:139]: to array is necessary? Can you use object in collection ?

File: cartridges/int_alma/cartridge/controllers/Product.js [hunk 5]
```diff
@@ -19,16 +19,20 @@ server.append('Show', function (req, res, next) {
...
     var almaPlanHelper = require('*/cartridge/scripts/helpers/almaPlanHelper');
+    var almaHelpers = require('*/cartridge/scripts/helpers/almaHelpers');
 
     var params = {
         method: 'GET',
         pid: req.querystring.pid
     };
     var product = ProductFactory.get(params);
+
+    var isWidgetEnabled = almaHelpers.haveExcludedCategory([req.querystring.pid]) ? false : almaWidgetHelper.isWidgetEnabled('product');
```
Comment (Francois-Gomis) [inline cartridges/int_alma/cartridge/controllers/Product.js:30]: Can you create a variable with : `req.querystring.pid`

File: cartridges/int_alma/cartridge/scripts/helpers/almaHelpers.js [hunk 6]
```diff
@@ -123,6 +123,43 @@ function getLocale(req) {
...
+                categoriesID.push(category.getID());
+            });
+        } else {
+            product.getMasterProduct().getAllCategories().toArray().forEach(function (category) {
```
Comment (Francois-Gomis) [inline cartridges/int_alma/cartridge/scripts/helpers/almaHelpers.js:143]: to array is necessary? Can you use object in collection ?

**Review by Francois-Gomis — APPROVED** [review]

**Pull request merged**